flow/connectors/connelasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,204 @@
+		esCfg.APIKey = *config.ApiKey
+	}
+
+	esClient, err := elasticsearch.NewClient(*esCfg)
+	if err != nil {
+		return nil, fmt.Errorf("error creating elasticsearch connector: %w", err)
+	}
+	pgMetadata, err := metadataStore.NewPostgresMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ElasticsearchConnector{
+		PostgresMetadata: pgMetadata,
+		client:           esClient,
+		logger:           logger.LoggerFromCtx(ctx),
+	}, nil
+}
+
+func (esc *ElasticsearchConnector) ConnectionActive(ctx context.Context) error {
+	err := esc.client.DiscoverNodes()
+	if err != nil {
+		return fmt.Errorf("failed to check if elasticsearch peer is active: %w", err)
+	}
+	return nil
+}
+
+func (esc *ElasticsearchConnector) Close() error {
+	// stateless connector
+	return nil
+}
+
+func (esc *ElasticsearchConnector) SetupQRepMetadataTables(ctx context.Context,
+	config *protos.QRepConfig,
+) error {
+	return nil
+}
+
+func (esc *ElasticsearchConnector) SyncQRepRecords(ctx context.Context, config *protos.QRepConfig,
+	partition *protos.QRepPartition, stream *model.QRecordStream,
+) (int, error) {
+	startTime := time.Now()
+
+	schema := stream.Schema()
+
+	var bulkIndexErrors []error
+	var bulkIndexMutex sync.Mutex
+	var docId string
+	numRecords := 0
+	bulkIndexerHasShutdown := false
+
+	// -1 means use UUID, >=0 means column in the record
+	upsertColIndex := -1
+	// only support single upsert column for now
+	if config.WriteMode.WriteType == protos.QRepWriteType_QREP_WRITE_MODE_UPSERT &&
+		len(config.WriteMode.UpsertKeyColumns) == 1 {
+		for i, field := range schema.Fields {
+			if config.WriteMode.UpsertKeyColumns[0] == field.Name {
+				upsertColIndex = i
+			}
+		}
+	}
+
+	esBulkIndexer, err := esutil.NewBulkIndexer(esutil.BulkIndexerConfig{
+		Index:  config.DestinationTableIdentifier,
+		Client: esc.client,
+		// parallelism comes from the workflow design itself, no need for this
+		NumWorkers:    1,
+		FlushInterval: 10 * time.Second,
+	})
+	if err != nil {
+		esc.logger.Error("[es] failed to initialize bulk indexer", slog.Any("error", err))
+		return 0, fmt.Errorf("[es] failed to initialize bulk indexer: %w", err)
+	}
+	defer func() {
+		if !bulkIndexerHasShutdown {
+			err := esBulkIndexer.Close(context.Background())
+			if err != nil {
+				esc.logger.Error("[es] failed to close bulk indexer", slog.Any("error", err))
+			}
+		}
+	}()
+
+	for qRecord := range stream.Records {
+		qRecordJsonMap := make(map[string]any)
+
+		if upsertColIndex >= 0 {
+			docId = fmt.Sprintf("%v", qRecord[upsertColIndex].Value())
+		} else {
+			docId = uuid.New().String()
+		}
+		for i, field := range schema.Fields {
+			switch field.Type {
+			// JSON is stored as a string, fix that
+			case qvalue.QValueKindJSON:
+				qRecordJsonMap[field.Name] = json.RawMessage(shared.
+					UnsafeFastStringToReadOnlyBytes(qRecord[i].(qvalue.QValueJSON).Val))
+			default:
+				qRecordJsonMap[field.Name] = qRecord[i].Value()
+			}
```

Comment:
```suggestion
			switch r := qRecord[i].(type) {
			// JSON is stored as a string, fix that
			case qvalue.QValueJSON:
				qRecordJsonMap[field.Name] = json.RawMessage(shared.
					UnsafeFastStringToReadOnlyBytes(r.Val))
			default:
				qRecordJsonMap[field.Name] = r.Value()
			}
```
This properly handles `QValueNull`